core/consensus/aura/primitives/src/lib.rs

```diff
@@ -37,19 +39,22 @@ decl_runtime_apis! {
 		///
 		/// Dynamic slot duration may be supported in the future.
 		fn slot_duration() -> u64;
+
+		/// Construct a call to report the equivocation.
+		fn construct_equiv_report_call(proof: AuraEquivProof<<Block as BlockT>::Header, Signature>) -> Vec<u8>;
```

Comment:
style: no compact names. prefer `construct_equivocation_report_call` and `AuraEquivocationProof`

Reply:
I wonder if these `constructor` functions should be automatically generated using some kind of annotation in the `target` function.

Reply:
It would be cool to auto-generate them. I'm not sure what you mean by the `target` though

Reply:
The idea was to decorate the function using something like `#[create_constructor]`:
```
decl_module! {
	pub struct Module<T: Trait> for enum Call where origin: T::Origin { 
		#[create_constructor]
                fn report_equivocation(...) {
		}
	}
}
```
to avoid polluting the API declaration. That should implicitly implement the following runtime API call:
```
fn construct_report_equivocation_call(...) -> Vec<u8> {
    let report_call = Call::Aura(AuraCall::report_equivocation(proof));
    let extrinsic = UncheckedExtrinsic::new_unsigned(report_call);
    extrinsic.encode()
}
```
But now I don't like so much the idea of creating *implicit* API functions. (By `implicit` I mean that doesn't appear under `decl_runtime_apis!`)
 
Maybe @bkchr has an opinion on this.

Reply:
I would really like to have implicit API functions, but this does not work. You can not export a function that involves some generics.